cpp/src/fil/internal.cuh

```diff
@@ -266,6 +302,90 @@ struct forest_params_t {
 /// FIL_TPB is the number of threads per block to use with FIL kernels
 const int FIL_TPB = 256;
 
+static const uint32_t max_precise_int_float = 1 << 24;  // 16'777'216
+
+struct cat_feature {
+  int max_matching = -1, n_nodes = 0;
+  void reduce_with(cat_feature b)
+  {
+    max_matching = std::max(max_matching, b.max_matching);
+    n_nodes += b.n_nodes;
+  }
+};
+typedef std::vector<cat_feature> v_cat_feature;
+struct categorical_branches {
```

Comment:
`categorical_branches` is definitely not the best name for this structure. Based on what it stores, I'd name it `categorical_bitmaps` or `categorical_sets`.

The same applies to variables of this type throughout the code.